Revolve improvements

zenfiric opened the pull request:
This batch of improvements contain:
- [x] Enable configuration options to users
- [x] Create brain selector

Commit:
Rename `brain` to `brains` directory

We plan to add several different learners/controllers,
so it is more accurate.

Commit:
Bundle different brain implementations under one header

Commit:
Add learner selector as a Python argument

Robots should be tested for different learners/controllers.
Adding a selector in the configurations makes this easier.

Commit:
Add brain selector in C++ plugin

Commit:
Add default `output` and `restore` directories

Commit:
Enable Python manager to parse user-settings

Commit:
Add an option to transfer robot's genome within SDF

Commit:
Rename some of frequently used method to become more understandable

Commit:
Change `++` from postfix to prefix form

Always prefix the `++` operator, which is slightly
more efficient than postfix in some cases.

Commit:
Replace Boost with STL smart pointers

Commit:
Add search for `yaml-cpp` for parsing YAML

Commit:
Add `YamlBodyParser` as a utility module for plugins

Commit:
Remove default constructor in `YamlBodyParser`

Commit:
Rename `yaml_to_robot` -> `yaml_to_proto` for clarity

portaloffreedom commented on revolve.py:
```diff
@@ -36,10 +36,10 @@ def write_stderr(self, data):
     supervisor = OnlineEvolutionSupervisor(
-        manager_cmd='python3',
-        # manager_args=['-u', os.path.join(here, "experiments/examples/manager.py")],
-        manager_args=['-u', settings.manager],
```
`-u` option is important or you don't have output from the manager

Commit:
Move `node_` to the base `Brain` class

Commit:
Some formatting of SDFBuilder

Commit:
Rename variable in `NeuralNetwork` for clarity

Commit:
Rename Protobuf `BodyPart` attributes for clarity

The attributes were named `src` and `dst`. This can lead
to a confusion for new developers. They are designating
IDs of connecting slots, and it is now more clear in the code.

Commit:
Rename more methods for clarity

Commit:
Add parts coordinates in Protobuf robots

Now x,y-coordinates are determined while parsing
the model and they can be further transferred to
SDF robot and eventually to RobotController plugin.

Commit:
Add selection for Differential CPG controller

Commit:
Fix Python tree representation to support coordinates

Commit:
Implement coordinates in SDF robot

We need coordinates for motor's positions, so we
only add (x, y) to `rv:motor` tag.

portaloffreedom commented on cpprevolve/revolve/gazebo/Types.h:
```diff
@@ -22,9 +22,6 @@
 
 #include <vector>
 
```
`#include <memory>`

zenfiric replied:
Added in #292cd7a commit.

portaloffreedom commented on cpprevolve/revolve/gazebo/brains/Brain.h:
```diff
@@ -12,9 +12,9 @@
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
-* Description: Specifies a utility `Brain` base class. If your brain doesn't
```
fix documentation

zenfiric replied:
Fixed in daebd2d

portaloffreedom commented on cpprevolve/revolve/gazebo/brains/NeuralNetwork.cpp:
```diff
@@ -105,15 +105,15 @@ NeuralNetwork::NeuralNetwork(
     {
       std::cerr << "Missing required neuron attributes (id or layer). '"
                 << std::endl;
-      throw std::runtime_error("Robot brain error");
```
brainsssssszz

zenfiric replied:
Fixed in daebd2d

portaloffreedom commented on cpprevolve/revolve/gazebo/brains/NeuralNetwork.cpp:
```diff
@@ -253,22 +253,22 @@ NeuralNetwork::NeuralNetwork(
     neuronHelper(&params_[pos * MAX_NEURON_PARAMS],
                  &types_[pos],
                  neuronMap[neuronId]);
-    outPos++;
```
`++outputIndex`

zenfiric replied:
Fix in 048524f

portaloffreedom commented on cpprevolve/revolve/gazebo/plugin/RobotController.cpp:
```diff
@@ -191,7 +190,19 @@ void RobotController::LoadBrain(const sdf::ElementPtr _sdf)
+  else if ("diff_cpg" == learner)
+  {
+    brain_.reset(new DifferentialCPG(this->model_, brain, motors_, sensors_));
+  }
```
else error

zenfiric replied:
Fixed in 05471c4

portaloffreedom commented on tools/compile_python:
```diff
@@ -1,9 +1,9 @@
 #!/bin/sh
```
rename file compile_python.sh

zenfiric replied:
Fixed in 53f9fef

Commit:
Added missing header

Commit:
Fix incorrect renaming `brain` -> `brains`

Commit:
Use prefix unary operators

Commit:
Throw an error if no brain is specified

Commit:
Rename `compile_python` -> `proto2python.sh`
